homeassistant/components/ipp/__init__.py

```diff
@@ -0,0 +1,200 @@
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+from homeassistant.helpers.entity import Entity
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+
+from .const import (
+    ATTR_IDENTIFIERS,
+    ATTR_MANUFACTURER,
+    ATTR_MODEL,
+    ATTR_SOFTWARE_VERSION,
+    CONF_BASE_PATH,
+    DOMAIN,
+)
+
+PLATFORMS = [SENSOR_DOMAIN]
+SCAN_INTERVAL = timedelta(seconds=60)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup(hass: HomeAssistant, config: Dict) -> bool:
+    """Set up the IPP component."""
+    hass.data.setdefault(DOMAIN, {})
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up IPP from a config entry."""
+
+    # Create IPP instance for this entry
+    coordinator = IPPDataUpdateCoordinator(
+        hass,
+        host=entry.data[CONF_HOST],
+        port=entry.data[CONF_PORT],
+        base_path=entry.data[CONF_BASE_PATH],
+        tls=entry.data[CONF_SSL],
+        verify_ssl=entry.data[CONF_VERIFY_SSL],
+    )
+    await coordinator.async_refresh()
+
+    if not coordinator.last_update_success:
+        raise ConfigEntryNotReady
+
+    hass.data[DOMAIN][entry.entry_id] = coordinator
+
+    # For backwards compat, set unique ID
+    if entry.unique_id is None:
+        hass.config_entries.async_update_entry(
+            entry, unique_id=coordinator.data.info.uuid
+        )
+
+    for component in PLATFORMS:
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(entry, component)
+        )
+
+    return True
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Unload a config entry."""
+    unload_ok = all(
+        await asyncio.gather(
+            *[
+                hass.config_entries.async_forward_entry_unload(entry, component)
+                for component in PLATFORMS
+            ]
+        )
+    )
+
+    if unload_ok:
+        hass.data[DOMAIN].pop(entry.entry_id)
+
+    return unload_ok
+
+
+class IPPDataUpdateCoordinator(DataUpdateCoordinator):
+    """Class to manage fetching IPP data from single endpoint."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        *,
+        host: str,
+        port: int,
+        base_path: str,
+        tls: bool,
+        verify_ssl: bool,
+    ):
+        """Initialize global IPP data updater."""
+        self.ipp = IPP(
+            host=host,
+            port=port,
+            base_path=base_path,
+            tls=tls,
+            verify_ssl=verify_ssl,
+            session=async_get_clientsession(hass, verify_ssl),
+        )
+
+        super().__init__(
+            hass, _LOGGER, name=DOMAIN, update_interval=SCAN_INTERVAL,
+        )
+
+    async def _async_update_data(self) -> IPPPrinter:
+        """Fetch data from IPP."""
+        try:
+            return await self.ipp.printer()
```

Comment:
Does this need to have a timeout?

Reply:
the underlying library handles timeouts.

https://github.com/ctalkington/python-ipp/blob/master/pyipp/ipp.py#L110